### PR TITLE
CoreDNS Governance proposition

### PR DESCRIPTION
**Pull request by fturib**
### Why is this pull request needed and what does it do?
In order to graduate on last level inside CNCF organization, we need to clarify our GOVERNANCE rules.

### Description

After MAINTAINERS sync-up of today, 

I merged the [proposition of Miek](https://github.com/coredns/coredns/pull/2152), 
replacing the chapter `Benevolent Dictator for Life` by a `Decision making process` moved toward  the end of the document.

This Decision making process is based on consensus at first, and, in last resort, a vote where each maintainer has one vote whatever the organization. (Miek has 2 votes).

There is no mention of organization in this document, it is purely based on maintainer whatever their professional status.

### CODE REVIEW

File: GOVERNANCE.md
Comment (dilyevsky) [inline GOVERNANCE.md]: Working link is https://github.com/coredns/coredns/blob/master/OWNERS

**Review by dilyevsky — COMMENTED** [review]
seems reasonable to me

**Review by bradbeam — APPROVED** [review]

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]: Is this a quote?  If so, it should be attributed.  If not, it should not have quotations around it.

Also as a general note, if any sections of this document are copied/borrowed, we should acknowledge the source.

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
file, with their Github handle and a possibly obfuscated email address. Everyone in the
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
"Everyone carries water."
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]: What does it mean to be "inactive"? What are the implications?
If there are none, we should remove this.

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
On successful merge of a significant pull request, any current maintainer can reach
```

Do we need to define "significant" here?

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
to the author and ask them if they are willing to become a CoreDNS
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
fails to adhere to the [Code of Conduct](https://github.com/coredns/coredns/blob/master/CODE-OF-CONDUCT.md), or meet the principles of the CoreDNS community as defined in this document.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
The CoreDNS organization is open to receive new sub-projects under its umbrella. To accept a project
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
into the __CoreDNS__ organization, it has to meet the following criteria:
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
- MUST be licensed under the terms of the Apache License v2.0.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
- MUST be related to one or more scopes of the CoreDNS ecosystem:
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
- MUST be supported by a Maintainer.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
  - Other DNS related processing
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md:60]: What precisely does it mean to be a CNCF sub-project under the umbrella of CoreDNS?

The project transfers governance to CoreDNS maintainers?

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
Proposals and ideas can either be submitted for agreement via an github issue or by sending an email to `maintainer@coredns.io`
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
For formal votes, the following should be added to the relevant github issue or PR:
* A specific statement of what is being voted on.
* The voting period - a suitable amount of time during which voting will occur.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
A link to that issue
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
or PR should be sent to `maintainer@coredns.io`.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
After the voting period is over, the votes will be tallied and the outcome noted.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
Maintainers should indicate their yes/no vote (or respectively +1/-1) on that issue or PR.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
A 2/3 majority of cast votes is needed for the statement to be approved.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]: It's unclear to me what section below this is referring to.

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
## Becoming a Maintainer
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
Making a community work requires input/effort from everyone. Maintainers should actively
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
The [CoreDNS Code of Conduct](https://github.com/coredns/coredns/blob/master/CODE-OF-CONDUCT.md) is aligned with the CNCF Code of Conduct.
```

File: GOVERNANCE.md
Comment (chrisohaver) [inline GOVERNANCE.md]:
```suggestion
Miek Gieben (@miekg), as the owner of CoreDNS, weighs two votes.
```

**Review by chrisohaver — CHANGES_REQUESTED** [review]

**Comment by chrisohaver** [comment]
What happened to the "corporate takeover prevention clause"?  Where maintainers working on behalf of an organization only get one vote collectively for their whole organization.

Are we not doing that anymore? Or is it in there and i'm just not seeing it?

File: GOVERNANCE.md
Comment (rajansandeep) [inline GOVERNANCE.md]:
```suggestion
`approvers` list is a Maintainer.
```

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: Thanks. I think I can simplify to OWNERS

**Comment by fturib** [comment]
> What happened to the "corporate takeover prevention clause"? Where maintainers working on behalf of an organization only get one vote collectively for their whole organization.


@miekg : I removed the initial limitation of a single vote per organization (which is the "Corporate takeover prevention clause" that @chrisohaver is referring above). 

I understood here that you prefer to have no reference at all to organization. 

But I may have misunderstood and be wrong here. 

**what is your position ?**




NOTE: Initial "corporate takeover prevention clause" is:


>The CoreDNS project employs "organization voting" to ensure no single organization can dominate the project.

>Individuals declare if they contribute as un-affiliated or as associated with or employee of an organization or a company.

>Individuals that contribute as un-affiliated are allowed one organization vote. Each company or organization (regardless of the number of maintainers associated with or employed by that company/organization) receives one organization vote.

>In other words, if two maintainers are employed by Company X, two by Company Y, two by Company Z, and one maintainer is an un-affiliated individual, a total of four "organization votes" are possible; one for X, one for Y, one for Z, and one for the un-affiliated individual.

>Any maintainer from an organization may cast the vote for that organization.

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: I agree. removed.

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: I do not think so. Anyway the idea is to welcome anyone that is interested to contribute.

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: right. will rather add a link to coredns.io/community that advertize it.

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: good catch.

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: hum. I do not think it is a quote.
I cannot find another match in google.

Comment (fturib) [reply GOVERNANCE.md]: about `Everyone carries water` that is more an idiom no ?

about sections copied .. yes some are, at least some sentences that I copy/paste from other cncf governance doc, which themselves seems copied. Are you sure we need to add the source here ? (I do not remind where I took it).

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: I am not sure how to define the owner of coreDNS - and if there is one, since we moved to cncf and it is opensource project. 
From point of view of Github there are several owners.

That is why I prefer the "historical owner" that show, whatever owners are defined today, that the project were started and grown up by Miek.

I propose to keep "historical owner" for now.

Comment (fturib) [reply GOVERNANCE.md]: unless there is a more appropriate way to say the same thing ?

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md]: for now, will remove the quotes.

**Comment by fturib** [comment]
@chrisohaver, @rajansandeep  : I think I either integrated your proposition of change or reply on your comment in this PR. Thanks to verify.

**Comment by chrisohaver** [comment]
> @chrisohaver, @rajansandeep : I think I either integrated your proposition of change or reply on your comment in this PR. Thanks to verify.

@fturib, It appears that most of my suggestions were not accepted ... Some are grammatical/spelling corrections that I would urge us to address before merging.

**Comment by fturib** [comment]
>Some are grammatical/spelling corrections that I would urge us to address before merging.

I though I addressed all those. Let me double check...
My bad ! ... I was looking into "conversation" that was showing only a part of your comments.
=> I will work integrating the propositions I did not see.

**Comment by fturib** [comment]
@chrisohaver : can you have a second look ? I think this time I addressed all of your comment (hopefully).

**Comment by chrisohaver** [comment]
Thanks @fturib,

About the "CNCF umbrella" thing, I think we should be prepared to be able to answer what that means if we are asked.  It would be even better to describe what it means in this document or link to a place that explains it.

**Review by chrisohaver — APPROVED** [review]

**Comment by fturib** [comment]
@miekg : can you review and provide your feedback or approval ?

File: GOVERNANCE.md
Comment (yongtang) [inline GOVERNANCE.md:73]: I noticed that organization vote restriction has been removed. It is fine as long as the organization stays with the principal of ~`open` and `transparent`~ `transparent and accessible`. I do think some clarifications should be made to make sure no single organization is able to dominate the project. I will share some of my thinking about this area.

**Review by yongtang — CHANGES_REQUESTED** [review]

**Comment by yongtang** [comment]
I think `transparent and accessible` is very important to prevent single organization from being able to dominate. Some clarifications might be needed here.

One important thing is that, make sure `decision making process` stays public, either through `maintainers@coredns.io` (open to every maintainers), or through pull requests on GitHub (open to every community members).

**Comment by yongtang** [comment]
I believe every coredns community members have the intentions to help the project better, so as long as everything remains public, different opinions will always be sorted out in a good way.

**Comment by yongtang** [comment]
We do need to make sure there is no `small circle decisions` by `a carefully selected group`.

There are several ways to get around the proposed `decision making process` and play the systems. For example:
- One maintainer or one organization could host a maintainer's meeting, invite only a small set of maintainers it likes (or invites many non-maintainers to water down the votes), and reaches a `self-claimed consensus`.
- Even if a meeting is open and accessible to every maintainers, the host of the meeting, could still `run the table` and sway the decisions by predefining the agenda.

How to make sure the above scenarios are not going to happen? I think 
- we have to make sure `maintainers@coredns.io` (open to all maintainers) or GitHub pull requests (open to everyone) are the only proper channels for any decision making process.
- any proposal should leave enough _reasonable_ time for every maintainer to think, digest, and express his or her opinions.

**Comment by yongtang** [comment]
Now coredns is under the umbrella of cncf. There will be communications between coredns project and cncf.

A question is, who should be the person of contact for coredns to cncf? Should the person be ~selected~ _elected_ by maintainers, or should `maintainers@coredns.io` be used for conversations between coredns and cncf?

I think `maintainers@coredns.io` could be a better way as it allows _every_ maintainer to be in the loop, though I am happy to hear any suggestions.

**Comment by yongtang** [comment]
Another question is, there are two proposals #2152 and #2203. Which one should be used as the basis for update and modification? and _why_?

File: GOVERNANCE.md
Comment (fturib) [reply GOVERNANCE.md:73]: >  I will share some of my thinking about this area.

Let me know what you would like to change here to address "no single organization is able to dominate the project". 

There is one in this [comment](https://github.com/coredns/coredns/pull/2203#issuecomment-430793776), that is coming from former version of the proposal, and that I copied from other project that had the same concern.

I would like also the opinion of @miekg on this, as he was reluctant to add "organization" reference in this document.

**Comment by fturib** [comment]
> How to make sure the above scenarios are not going to happen? 

I am not sure you can avoid that any side-conversation between maintainers or non-maintainers can happen.
But from the 'making decision process', there is always a way for any maintainer to verify that decision is common: it is by calling the vote. And this mechanism seems clear and public.

**Comment by fturib** [comment]
> who should be the person of contact for coredns to cncf? 

Here I am not sure as there is no person dedicated for that.

CNCF is part of this project, so let's ask @caniszczyk to tell us how it is defined for other projects under umbrella of cncf.

**Comment by fturib** [comment]
> Another question is, there are two proposals #2152 and #2203. Which one should be used as the basis for update and modification? and why?

Hum ... initially I created one proposal : #2112, then Miek made another proposal #2152. 
After sync-up I closed the initial #2112, and created this one, based on #2152, as I explained in the [description of this PR](https://github.com/coredns/coredns/pull/2203#issue-223418913).

Looks like the whole point is too agree on the chapter `Decision making process` initially `Benevolent Dictator for Life`

We can work on any of these PR that have the same root.
If you prefer to update #2152, up to you, that is fine with me.
(In that case, please ensure to take care about all grammatical/spelling change called by @chrisohaver that I have already addressed in this PR).

**Comment by yongtang** [comment]
> How to make sure the above scenarios are not going to happen?
> I am not sure you can avoid that any side-conversation between maintainers or non-maintainers can happen.
> But from the 'making decision process', there is always a way for any maintainer to verify that decision is common: it is by calling the vote. And this mechanism seems clear and public.

Any maintainer can talk to any one, any maintainer can also have sync up meetings with any one, or any group of his or her choice.

The clarification is make sure any "decision" is coming from `maintainers@coredns.io` or GitHub pull requests, not from "oh, two or three of us met, and here are the decision". Keep in mind the proposed `Decision making process` have the following clause:
```
Proposals and ideas can either be submitted for agreement via an \
github issue or by sending an email to `maintainer@coredns.io`
```
If decisions from other means are ok, you can certainly modify the proposed `Decision making process`.

**Comment by fturib** [comment]
>Keep in mind the proposed Decision making process have the following clause:
>`Proposals and ideas can either be submitted for agreement via an \
github issue or by sending an email to `maintainer@coredns.io`
> If decisions from other means are ok, you can certainly modify the proposed Decision making process.

I am not sure how to interpret ...   <br>
I understand that you agree that the current clause in `Decision making process` is good enough and ensure decision will be public. So we do not need to change.

Am I correct ?

**Comment by yongtang** [comment]
The clause is certainly good enough, expect what are the implications if it is not followed?

Similarly, some other comments from the review:
> What does it mean to be "inactive"? What are the implications?
> If there are none, we should remove this.

**Comment by yongtang** [comment]
I would prefer the discussion to be in the other PR, but just in case, here is the copied context https://github.com/coredns/coredns/pull/2152#issuecomment-431593280:

I think the issue with coredns is that, coredns is not unique compared with many other open source projects. It is however, might be special inside cncf.

Many other open source projects in cncf have a (or more than one) sponsoring company and the company behind is also the major technical contributor. Companies certainly have more non-technical resources but that typically aligns with technical contributions collectively.

So cncf projects still largely follows merit-based open source model and works well.

For coredns, there is one independent contributor (@miekg) that has more than 50% of the technical contributions day-to-day. There is also a company that may have more non-technical resources, but, simply just could not have enough technical contributions.

Open source software should be merit-based in the area of technical contributions.

For me, it will be a _bigger_ joke if _THE_ major contributor of the project is not involved or does not have a final say on anything in this project.

Note things may change in the future.

**Comment by yongtang** [comment]
Continue https://github.com/coredns/coredns/pull/2152#issuecomment-431593713:

A merit-based `project lead` can help incubating, set vision, and reduce noice early on. Things may change when contributions of a project diversify.

But at the current stage, I see more benefits for a `project lead` that drives the direction of the coredns.

**Comment by fturib** [comment]
>A merit-based project lead can help incubating, set vision, and reduce noice early on. Things may change when contributions of a project diversify.

>But at the current stage, I see more benefits for a project lead that drives the direction of the coredns.

I think that can be fine to have a project lead, as long as we define its role, how that project lead is elected, for how long, and what is the process of replacement/continuation. 
I mean a periodic revision of who can take this role - based on merit -, with the agreement of the CoreDNS community.

That would made an "open governance" - because of the periodic validation by the community - and keep all the positive point to have a quick decision process. 

what do you think ?

**Comment by fturib** [comment]
Closing this proposition.
The active proposition for Governance is here : #2220

**Pull request closed without merging**
